src/IECoreScene/ShaderNetworkAlgo.cpp

```diff
@@ -233,70 +228,245 @@ void ShaderNetworkAlgo::convertOSLComponentConnections( ShaderNetwork *network,
+
+		toRemove.push_back( s.first );
+
+		ShaderNetwork::ConnectionRange inputConnections = network->inputConnections( s.first );
+		ShaderNetwork::ConnectionRange outputConnections = network->outputConnections( s.first );
 
-					for( int i = 0; i < 3; ++i )
+		for( ShaderNetwork::ConnectionIterator it = outputConnections.begin(); it != outputConnections.end(); )
+		{
+			// Copy and increment now so we still have a valid iterator
+			// if we remove the connection.
+			const ShaderNetwork::Connection connection = *it++;
+			network->removeConnection( connection );
+
+			if( isPack )
+			{
+				const Shader *targetShader = network->getShader( connection.destination.shader );
+
+				for( ShaderNetwork::ConnectionIterator inputIt = inputConnections.begin(); inputIt != inputConnections.end(); inputIt++ )
+				{
+					const IECore::InternedString &inputName = inputIt->destination.name;
+					int inputIndex = -1;
+					for( int i = 0; i < 3; i++ )
 					{
-						ShaderNetwork::Parameter source = network->input( { shader.first, parameterName.string() + "." + g_vectorComponents[i] } );
-						if( !source )
+						if( inputName == g_packInParameterNames[i] )
 						{
-							source = network->input( { shader.first, parameterName.string() + "." + g_colorComponents[i] } );
-						}
-						if( source )
-						{
-							network->addConnection( { source, { packHandle, g_packInParameterNames[i] } } );
+							inputIndex = i;
 						}
 					}
 
-					network->removeConnection( connection );
+					if( inputIndex == -1 )
+					{
+						throw IECore::Exception( boost::str(
+							boost::format(
+								"removeComponentConnectionAdapters : Unrecognized input for mx_pack_color \"%1%\""
+							) % inputName
+						) );
+					}
+
+					ShaderNetwork::Parameter componentDest;
+					if( targetShader->parametersData()->member<Color3fData>( connection.destination.name ) )
+					{
+						componentDest = { connection.destination.shader, IECore::InternedString( connection.destination.name.string() + "." + g_colorComponents[inputIndex] ) };
+					}
+					else if( targetShader->parametersData()->member<V3fData>( connection.destination.name ) )
+					{
+						componentDest = { connection.destination.shader, IECore::InternedString( connection.destination.name.string() + "." + g_vectorComponents[inputIndex] ) };
+					}
+					else
+					{
+						throw IECore::Exception( boost::str(
+							boost::format(
+								"removeComponentConnectionAdapters : Unrecognized type for targe parameter \"%1%.%2%\""
+							) % connection.destination.shader.string() % connection.destination.name.string()
+						) );
+					}
+
+					network->addConnection( { inputIt->source, componentDest } );
+				}
+			}
+			else
+			{
+				if( inputConnections.begin() == inputConnections.end() )
+				{
+					throw IECore::Exception( boost::str(
+						boost::format(
+							"removeComponentConnectionAdapters : mx_swizzle_color_float \"%1%\" must have an input"
+						) % s.first.string()
+					) );
 				}
```

Comment:
This assumes that if there's at least one connection, the first one must be into `swizzle.in`. We could check more accurately by calling `ShaderNetwork::input()`, and then we wouldn't need `inputConnections` at all for this branch.